Author: Luca Boccassi <luca.boccassi@gmail.com>
Description: Undef CONFIG_XEN and CONFIG_XEN_DOM0 if building on ARM
 Currently armmp kernel headers (3.16, 4.0, 4.1) have CONFIG_XEN enabled and
 CONFIG_PARAVIRT disabled. This causes a build failure of the armhf dkms modules
 in all current versions of the Nvidia driver.
 Since running this driver on Xen is currently not supported anyway on Debian,
 this patch undefines CONFIG_XEN and CONFIG_XEN_DOM0 if building on ARM, and
 avoid including asm/maddr.h and xen/interface/memory.h.
 Note that this has been tested only on qemu.
 It is recommended to build an armmp kernel with CONFIG_XEN disabled (or with
 CONFIG_XEN_PARAVIRT enabled) to use the Nvidia drivers.

--- a/common/inc/nv-linux.h
+++ b/common/inc/nv-linux.h
@@ -42,5 +42,10 @@
 #include <linux/utsname.h>
 
+#if defined(NVCPU_ARM) && LINUX_VERSION_CODE < KERNEL_VERSION(3,17,0)
+#undef CONFIG_XEN
+#undef CONFIG_XEN_DOM0
+#endif
+
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 9)
 #error "This driver does not support kernels older than 2.6.9!"
 #elif LINUX_VERSION_CODE < KERNEL_VERSION(2, 7, 0)
@@ -258,7 +263,7 @@ NV_STATUS nvos_forward_error_to_cray(str
  * NVIDIA graphics driver has no way of knowing and is unlikely
  * to work correctly.
  */
-#if defined(CONFIG_XEN) && !defined(CONFIG_PARAVIRT)
+#if defined(CONFIG_XEN) && !defined(CONFIG_PARAVIRT) && !defined(NVCPU_ARM)
 #include <asm/maddr.h>
 #include <xen/interface/memory.h>
 #define NV_XEN_SUPPORT_FULLY_VIRTUALIZED_KERNEL
--- a/conftest.sh
+++ b/conftest.sh
@@ -39,7 +39,8 @@
     # CONFIG_XEN and CONFIG_PARAVIRT are present, text_xen() treats
     # the kernel as a stand-alone kernel.
     #
-    if ! test_configuration_option CONFIG_XEN ||
+    if [ "$ARCH" = "arm" ] ||
+         ! test_configuration_option CONFIG_XEN ||
          test_configuration_option CONFIG_PARAVIRT; then
         XEN_PRESENT=0
     fi
